worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/OffsetsMask2D.java

```diff
@@ -0,0 +1,180 @@
+ */
+
+package com.sk89q.worldedit.function.mask;
+
+import com.google.common.collect.ImmutableList;
+import com.sk89q.worldedit.math.BlockVector2;
+import com.sk89q.worldedit.math.BlockVector3;
+import com.sk89q.worldedit.util.Direction;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Checks whether any face of the given offsets to a block match a given mask.
+ */
+public class OffsetsMask2D extends AbstractMask2D {
+
+    private static final List<BlockVector2> OFFSET_LIST = ImmutableList.copyOf(
+        Direction.valuesOf(Direction.Flag.CARDINAL)
+            .stream()
+            .map(Direction::toBlockVector)
+            .map(BlockVector3::toBlockVector2)
+            .collect(Collectors.toList())
+    );
+
+    private Mask2D mask;
+    private final boolean excludeSelf;
+    private final int minMatches;
+    private final int maxMatches;
+    private final List<BlockVector2> offsets;
+
+    /**
+     * Create a new instance.
+     *
+     * @param mask the mask to test against
+     */
+    public OffsetsMask2D(Mask2D mask) {
+        this(mask, false);
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param mask the mask to test against
+     * @param excludeSelf excludes blocks where the mask matches itself
+     */
+    public OffsetsMask2D(Mask2D mask, boolean excludeSelf) {
+        this(mask, excludeSelf, 1, Integer.MAX_VALUE);
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param mask the mask to test against
+     * @param excludeSelf excludes blocks where the mask matches itself
+     * @param minMatches the minimum number of matches
+     * @param maxMatches the maximum number of matches
+     */
+    public OffsetsMask2D(Mask2D mask, boolean excludeSelf, int minMatches, int maxMatches) {
+        this(mask, excludeSelf, minMatches, maxMatches, OFFSET_LIST);
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param mask the mask to test against
+     * @param excludeSelf excludes blocks where the mask matches itself
+     * @param minMatches the minimum number of matches
+     * @param maxMatches the maximum number of matches
+     * @param offsets the block offsets to test with
+     */
+    public OffsetsMask2D(Mask2D mask, boolean excludeSelf, int minMatches, int maxMatches, List<BlockVector2> offsets) {
+        // Validate match args. No need to test maxMatches as it must be >=0 based on the conditions here.
+        checkArgument(minMatches <= maxMatches, "minMatches must be less than or equal to maxMatches");
+        checkArgument(minMatches >= 0, "minMatches must be greater than or equal to 0");
+        checkArgument(minMatches <= offsets.size(), "minMatches must be less than or equal to the number of offsets");
+        checkNotNull(mask);
+        checkNotNull(offsets);
```

Comment:
This is redundant by this point, should go at the top of the checks if you want to keep it.